apps/server/src/shared/domain/entity/ltitool.entity.ts

```diff
@@ -1,31 +1,129 @@
-import { Entity, Property } from '@mikro-orm/core';
-import { Optional } from '@nestjs/common';
+import { Entity, Enum, Property, Unique } from '@mikro-orm/core';
+import { ObjectId } from '@mikro-orm/mongodb';
+import { EntityId } from '@shared/domain/types';
 import { BaseEntityWithTimestamps } from './base.entity';
 
 export type ILtiToolProperties = Readonly<Omit<LtiTool, keyof BaseEntityWithTimestamps>>;
 
+export enum LtiRoleType {
+	LEARNER = 'Learner',
+	INSTRUCTOR = 'Instructor',
+	CONTENT_DEVELOPER = 'ContentDeveloper',
+	ADMINISTRATOR = 'Administrator',
+	MENTOR = 'Mentor',
+	TEACHING_ASSISTANT = 'TeachingAssistant',
+}
+
+export enum LtiPrivacyPermission {
+	ANONYMOUS = 'anonymous',
+	EMAIL = 'e-mail',
+	NAME = 'name',
+	PUBLIC = 'public',
+	PSEUDONYMOUS = 'pseudonymous',
+}
+
+export interface CustomLtiProperty {
+	key: string;
+	value: string;
+}
+
 @Entity({ tableName: 'ltitools' })
 export class LtiTool extends BaseEntityWithTimestamps {
-	@Property()
+	@Property({ nullable: false })
 	name: string;
 
-	@Property()
-	@Optional()
-	oAuthClientId?: string;
+	@Property({ nullable: false })
+	url: string;
+
+	@Property({ nullable: true })
+	key: string;
+
+	@Property({ nullable: false, default: 'none' })
+	secret: string;
+
+	@Property({ nullable: true })
+	logo_url?: string;
+
+	@Property({ nullable: true })
+	lti_message_type?: string;
+
+	@Property({ nullable: true })
+	lti_version?: string;
+
+	@Property({ nullable: true })
+	resource_link_id?: string;
+
+	@Enum({ array: true, items: () => LtiRoleType })
+	@Property({ nullable: false })
+	roles?: LtiRoleType[];
 
-	@Property()
-	@Optional()
-	secret?: string;
+	@Enum({
+		items: () => LtiPrivacyPermission,
+		default: LtiPrivacyPermission.ANONYMOUS,
+		nullable: false,
+	})
+	privacy_permission?: LtiPrivacyPermission;
 
-	@Property()
-	@Optional()
+	@Property({ nullable: false })
+	customs: CustomLtiProperty[];
+
+	@Property({ nullable: false, default: false })
+	isTemplate: boolean;
+
+	@Property({ nullable: true })
 	isLocal?: boolean;
 
+	@Property({ nullable: true, fieldName: 'originTool' })
+	_originToolId?: ObjectId;
+
+	// TODO tests
+	@Property({ persist: false, getter: true })
+	get originToolId(): EntityId | undefined {
+		return this._originToolId?.toHexString();
+	}
+
+	@Property({ nullable: true })
+	oAuthClientId?: string;
```

Comment:
why does an LTI-tool need an oauthClientId?

Reply:
its the old model has this property. we only mirror the model for porting the logic from feathers to nest